src/controller/python/chip/storage/delegate.py

```diff
@@ -0,0 +1,131 @@
+        self.lock = threading.Lock()
+        self.con = sqlite3.connect(options.path, check_same_thread=False)
+        self.cache = dict()
+        cur = self.con.cursor()
+        cur.executescript('''
+        CREATE TABLE IF NOT EXISTS storage (
+            domain text NOT NULL,
+            field text NOT NULL,
+            key text NOT NULL,
+            value text NOT NULL
+        );
+        CREATE UNIQUE INDEX IF NOT EXISTS storage_index ON storage (
+            domain, field, key
+        );
+        ''')
+        self.con.commit()
+
+    def Set(self, domain:str, field:str, key:str, value:bytes)->None:
+        with self.lock:
+            cur = self.con.cursor()
+            valueStr = base64.standard_b64encode(value).decode('utf-8')
+            cur.execute('INSERT OR REPLACE INTO storage (domain, field, key, value) VALUES (?, ?, ?, ?)', (domain, field, key, valueStr))
+            self.con.commit()
+
+    def Get(self, domain:str, field:str, key:str)->bytes:
+        with self.lock:
+            cur = self.con.cursor()
+            cur.execute("SELECT value FROM storage WHERE domain = ? AND field = ? AND key = ? LIMIT 1", (domain, field, key))
+            res = cur.fetchone()
+            return base64.standard_b64decode(res[0]) if res is not None else None
+
+    def Delete(self, domain:str, field:str, key:str)->None:
+        with self.lock:
+            cur = self.con.cursor()
+            cur.execute("DELETE FROM storage WHERE domain = ? AND field = ? AND key = ?", (domain, field, key))
+            self.con.commit()
+
+_controllerStorage = None
+
+@_ControllerGetKeyValueFunct
+def ControllerGetKeyValueFunct(key:c_char_p, keyLen:int, valPtr:c_void_p, sizePtr:ctypes.POINTER(c_uint16)):
+    global _controllerStorage
+    accessKey = ctypes.string_at(key, keyLen).decode("utf-8")
+    data = _controllerStorage.Get(_CONTROLLER_STORAGE_DOMAIN, _CONTROLLER_STORAGE_FIELD, accessKey)
+    sizeRead = ctypes.pointer(c_uint16(len(data))) if data is not None else ctypes.pointer(c_uint16(0xffff))
+    if data is not None and valPtr != 0:
+        copySize = min(len(data), sizePtr.contents.value)
+        ctypes.memmove(valPtr, ctypes.c_char_p(data), copySize)
+    ctypes.memmove(sizePtr, sizeRead, ctypes.sizeof(c_uint16))
+
+@_ControllerSetKeyValueFunct
+def ControllerSetKeyValueFunct(key:c_char_p, keyLen: int, valPtr:c_void_p, size:int):
+    global _controllerStorage
+    accessKey = ctypes.string_at(key, keyLen).decode("utf-8")
+    data = ctypes.string_at(valPtr, size)
+    _controllerStorage.Set(_CONTROLLER_STORAGE_DOMAIN, _CONTROLLER_STORAGE_FIELD, accessKey, data)
+
+@_ControllerDeleteKeyValueFunct
+def ControllerDeleteKeyValueFunct(key:c_char_p, keyLen: int):
+    global _controllerStorage
+    accessKey = ctypes.string_at(key, keyLen).decode("utf-8")
+    _controllerStorage.Delete(_CONTROLLER_STORAGE_DOMAIN, _CONTROLLER_STORAGE_FIELD, accessKey)
+
+def _InitControllerStorageInternal(storageClass, options: StorageOptions):
```

Comment:
these global methods and global access do not feel pythonic.

Could we at least have a storage class that exposes everything and eventually have one single 'global object' for which we call methods?

Python generally tries to do 

```python

storage = Storage(...)
storage.Init(...)  # although init should probably be done in the constructor
```